velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -151,25 +148,33 @@ class S3ReadFile final : public ReadFile {
   // bytes.
   void preadInternal(uint64_t offset, uint64_t length, char* position) const {
     // Read the desired range of bytes.
-    Aws::S3::Model::GetObjectRequest request;
-    Aws::S3::Model::GetObjectResult result;
-
-    request.SetBucket(awsString(bucket_));
-    request.SetKey(awsString(key_));
-    std::stringstream ss;
-    ss << "bytes=" << offset << "-" << offset + length - 1;
-    request.SetRange(awsString(ss.str()));
-    request.SetResponseStreamFactory(
-        AwsWriteableStreamFactory(position, length));
-    auto outcome = client_->GetObject(request);
-    VELOX_CHECK_AWS_OUTCOME(outcome, "Failed to get S3 object", bucket_, key_);
+    Aws::Transfer::TransferManagerConfiguration transfer_config(
+        executor_.get());
+    transfer_config.s3Client = client_;
+    auto transfer_manager =
+        Aws::Transfer::TransferManager::Create(transfer_config);
+    Aws::Utils::Stream::PreallocatedStreamBuf streamBuffer(
+        (unsigned char*)position, length);
+    auto downloadHandle =
+        transfer_manager->DownloadFile(bucket_, key_, offset, length, [&]() {
+          return Aws::New<UnderlyingStreamWrapper>("TestTag", &streamBuffer);
+        });
+    downloadHandle->WaitUntilFinished();
+    if (downloadHandle->GetStatus() != Aws::Transfer::TransferStatus::COMPLETED)
+      VELOX_FAIL(
+          "{} from location: {}:{}",
+          "Failed to get S3 object using the transfer manager",
+          bucket_,
+          key_);
   }
 
-  Aws::S3::S3Client* client_;
+  std::shared_ptr<Aws::S3::S3Client> client_;
   std::string bucket_;
   std::string key_;
   int64_t length_ = -1;
+  const std::shared_ptr<Aws::Utils::Threading::PooledThreadExecutor>& executor_;
```

Comment:
const ref to a shared ptr? I guess you would either use a const ref& to the object or copy the shared ptr here.

Reply:
Yeah, old habits from before std:*_ptrs were in heavy use.  Guess I am showing my age.  ;)